parsers/US_IPC.py

```diff
@@ -0,0 +1,145 @@
+#!usr/bin/env python3
+
+"""Parser for the Idaho Power Comapny area of the United States."""
+
+from dateutil import parser, tz
+from itertools import groupby
+from logging import getLogger
+import requests
+
+
+# NOTE No pumped storage yet but future ideas can be found at the following url.
+# https://docs.idahopower.com/pdfs/AboutUs/PlanningForFuture/irp/IRP.pdf
+
+PRODUCTION_URL = 'https://api.idahopower.com/Energy/Api/V1/GenerationAndDemand/Subset'
+
+GENERATION_MAPPING = {'Non-Utility Geothermal': 'geothermal',
+                      'Hydro': 'hydro',
+                      'Coal': 'coal',
+                      'Diesel': 'oil',
+                      'PURPA/Non-Utility Wind': 'wind',
+                      'Natural Gas': 'gas',
+                      'PURPA/Non-Utility Solar': 'solar',
+                      'PURPA Other': 'biomass'
+                      }
+
+
+def get_data(session=None):
+    """Returns a list of dictionaries."""
+
+    s = session or requests.Session()
+
+    req = requests.get(PRODUCTION_URL)
+    json_data = req.json()
+    raw_data = json_data['list']
+
+    return raw_data
+
+
+def timestamp_converter(timestamp):
+    """Converts a str timestamp into an aware datetime object."""
+
+    dt_naive = parser.parse(timestamp)
+    mountain_time = tz.gettz("America/Boise")
+    dt_aware = dt_naive.replace(tzinfo=dt_naive.tzinfo or mountain_time)
+
+    return dt_aware
+
+
+def data_processer(raw_data, logger):
+    """
+    Groups dictionaries by datetime key.
+    Removes unneeded keys and logs any new ones.
+    Returns a list of tuples containing (datetime object, dictionary).
+    """
+
+    dt_key = lambda x: x['datetime']
+    grouped = groupby(raw_data, dt_key)
+
+    unmapped = set()
+    parsed_data = []
+    for group in grouped:
+        dt = timestamp_converter(group[0])
+        generation = group[1]
+
+        production = {}
+        for gen_type in generation:
+            production[gen_type['name']] = float(gen_type['data'])
+
+        keys_to_remove = {'Load', 'Net Purchases', 'Inadvertent'}
+        known_keys = GENERATION_MAPPING.keys() | set() | keys_to_remove
```

Comment:
The result of `known_keys` here won't change from iteration to iteration. You could move the whole thing up outside the for loop, and have it be something like:

```
keys_to_ignore = {'Load', 'Net Purchases', 'Inadvertent'}
known_keys = GENERATION_MAPPING.keys() | keys_to_ignore
```

then in the loop, `keys_to_remove = keys_to_ignore | unknown_keys`